tokio-executor/src/task/core.rs

```diff
@@ -0,0 +1,158 @@
+use crate::loom::alloc::Track;
+use crate::loom::cell::CausalCell;
+use crate::task::Schedule;
+use crate::task::state::State;
+use crate::task::raw::{self, Vtable};
+use crate::task::waker::waker_ref;
+
+use std::cell::UnsafeCell;
+use std::future::Future;
+use std::mem::MaybeUninit;
+use std::pin::Pin;
+use std::ptr::{self, NonNull};
+use std::task::{Context, Poll, Waker};
+
+/// The task cell. Contains the components of the task.
+///
+/// It is critical for `Header` to be the first field as the task structure will
+/// be referenced by both *mut Cell and *mut Header.
+#[repr(C)]
+pub(super) struct Cell<T: Future, S: 'static> {
+    /// Hot task state data
+    pub(super) header: Header<S>,
+
+    /// Either the future or output, depending on the execution stage.
+    pub(super) core: Core<T>,
+
+    /// Cold data
+    pub(super) trailer: Trailer,
+}
+
+/// The core of the task.
+///
+/// Holds the future or output, depending on the stage of execution.
+pub(super) struct Core<T: Future> {
```

Comment:
Is there a reason for adding this struct as a wrapper around `Stage`, rather than just using `Stage` directly? I'm assuming it's not due to compatibility assumptions about enums, as this type is a private API?

Reply:
Fits better IMO. Encapsulate the "owned" data and not expose the enum variants.